presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -878,6 +878,27 @@ private boolean canRunMore()
         }
     }
 
+    // update eligibility for all nodes in the subtree rooted at this node
+    protected void updateEligibilityAll()
```

Comment:
I'd favor adding a "recursive" flag to updateEligibility instead of creating a new method.

Also, check to see if lines 776-779 can use this instead

Reply:
updateEligibility is also a recursive method, that goes up the tree from a node. This one goes down. So a parameter for direction? Another option is  separating the node-specific logic for updating eligibility to a third (non-recursive) method, that only updates eligibility for a node. These two recursive methods will then invoke the same new updateEligibility (pertaining to only one node) method in the manner they desire.